tests/test_linops/test_linops.py

```diff
@@ -0,0 +1,277 @@
+    assert linop_matvec.dtype == matrix_matvec.dtype
+
+    np.testing.assert_allclose(linop_matvec, matrix_matvec)
+
+
+@pytest.mark.parametrize("n", [1, 2, 15, 100])
+@pytest_cases.parametrize_with_cases("linop,matrix", cases=case_modules)
+def test_rmatmat(
+    linop: pn.linops.LinearOperator,
+    matrix: np.ndarray,
+    random_state: np.random.RandomState,
+    n: int,
+):
+    mat = random_state.normal(size=(n, linop.shape[0]))
+
+    linop_matmat = np.conj(mat) @ linop
+    matrix_matmat = np.conj(mat) @ matrix
+
+    assert linop_matmat.ndim == 2
+    assert linop_matmat.shape == matrix_matmat.shape
+    assert linop_matmat.dtype == matrix_matmat.dtype
+
+    np.testing.assert_allclose(linop_matmat, matrix_matmat)
+
+
+@pytest_cases.parametrize_with_cases("linop,matrix", cases=case_modules)
+def test_rmatmat_shape_mismatch(
+    linop: pn.linops.LinearOperator,
+    matrix: np.ndarray,
+):
+    mat = np.zeros((10, linop.shape[0] + 1))
+
+    with pytest.raises(Exception) as excinfo:
+        mat @ matrix  # pylint: disable=pointless-statement
+
+    with pytest.raises(excinfo.type):
+        mat @ linop  # pylint: disable=pointless-statement
+
+    mat = np.zeros((1, 10, linop.shape[0]))
+
+    with pytest.raises(ValueError):
+        mat[None, :, :] @ linop  # pylint: disable=pointless-statement
+
+
+@pytest_cases.parametrize_with_cases("linop,matrix", cases=case_modules)
+def test_todense(linop: pn.linops.LinearOperator, matrix: np.ndarray):
+    linop_dense = linop.todense()
+
+    assert isinstance(linop_dense, np.ndarray)
+    assert linop_dense.shape == matrix.shape
+    assert linop_dense.dtype == matrix.dtype
+
+    np.testing.assert_allclose(linop_dense, matrix)
+
+
+@pytest_cases.parametrize_with_cases("linop,matrix", cases=case_modules)
+def test_rank(linop: pn.linops.LinearOperator, matrix: np.ndarray):
+    linop_rank = linop.rank()
+    matrix_rank = np.linalg.matrix_rank(matrix)
+
+    assert isinstance(linop_rank, np.intp)
+    assert linop_rank.shape == ()
+    assert linop_rank.dtype == matrix_rank.dtype
+
+    assert linop_rank == matrix_rank
+
+
+@pytest_cases.parametrize_with_cases("linop,matrix", cases=case_modules)
+def test_eigvals(linop: pn.linops.LinearOperator, matrix: np.ndarray):
+    if linop.is_square:
+        linop_eigvals = linop.eigvals()
+        matrix_eigvals = np.linalg.eigvals(matrix)
+
+        assert isinstance(linop_eigvals, np.ndarray)
+        assert linop_eigvals.shape == matrix_eigvals.shape
+        assert linop_eigvals.dtype == matrix_eigvals.dtype
+
+        np.testing.assert_allclose(linop_eigvals, matrix_eigvals)
+    else:
+        with pytest.raises(np.linalg.LinAlgError):
+            linop.eigvals()
+
+
+@pytest.mark.parametrize("p", [None, 2])
+@pytest_cases.parametrize_with_cases("linop,matrix", cases=case_modules)
+def test_cond(linop: pn.linops.LinearOperator, matrix: np.ndarray, p: Union[None, int]):
+    linop_det = linop.cond(p=p)
```

Comment:
naming should be `linop_cond` and `matrix_cond`. This seems like a copy-paste error.

Reply:
You got me there